File: android/app/src/main/java/com/algonit/algo/features/chat/data/model/ChatResponse.kt
package com.algonit.algo.features.chat.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Full API response from POST /chat/message and POST /chat/confirm.
 */
@Serializable
data class ChatAPIResponse(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("message_id") val messageId: String,
    val response: ChatResponseBody,
    val intent: IntentInfo? = null,
    val timestamp: String
) {
    /**
     * Converts the API response into a [Message] for use in the UI layer.
     */
    fun toMessage(): Message = Message(
        id = messageId,
        role = Message.ROLE_ASSISTANT,
        content = response.text,
        structuredData = response.structuredData,
        suggestedActions = response.suggestedActions,
        intent = intent?.classifiedAs,
        requiresConfirmation = response.requiresConfirmation,
        confirmationId = response.confirmationId,
        timestamp = timestamp
    )
}

/**
 * The response body within a [ChatAPIResponse], containing the assistant's
 * text reply and any structured data or suggested actions.
 */
@Serializable
data class ChatResponseBody(
    val text: String,
    @SerialName("structured_data") val structuredData: StructuredData? = null,
    @SerialName("suggested_actions") val suggestedActions: List<SuggestedAction>? = null,
    @SerialName("requires_confirmation") val requiresConfirmation: Boolean? = null,
    @SerialName("confirmation_id") val confirmationId: String? = null
)

/**
 * Intent classification metadata from the agent's reasoning.
 */
@Serializable
data class IntentInfo(
    @SerialName("classified_as") val classifiedAs: String,
    val confidence: Double
)

/**
 * Token budget information for rate limiting awareness.
 */
@Serializable
data class TokenBudgetInfo(
    @SerialName("tokens_used") val tokensUsed: Int,
    @SerialName("tokens_remaining") val tokensRemaining: Int,
    @SerialName("tokens_total") val tokensTotal: Int,
    @SerialName("reset_at") val resetAt: String? = null
)

// --- Request models ---

/**
 * Request body for POST /chat/message.
 */
@Serializable
data class ChatRequest(
    @SerialName("conversation_id") val conversationId: String? = null,
    val message: String,
    @SerialName("input_type") val inputType: String = "text",
    @SerialName("audio_url") val audioUrl: String? = null,
    val context: ChatContext? = null
)

/**
 * Optional context sent with a chat message to provide the assistant
 * with information about the user's current state.
 */
@Serializable
data class ChatContext(
    val screen: String? = null,
    @SerialName("selected_campaign_id") val selectedCampaignId: String? = null
)

/**
 * Request body for POST /chat/confirm.
 */
@Serializable
data class ConfirmRequest(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("confirmation_id") val confirmationId: String,
    val confirmed: Boolean
)

// --- Additional response models ---

/**
 * Response from GET /chat/suggestions, providing contextual suggestions
 * for the user to start a conversation.
 */
@Serializable
data class SuggestionsResponse(
    val suggestions: List<SuggestedAction>
)

/**
 * Response from GET /chat/budget, providing token usage information.
 */
@Serializable
data class BudgetResponse(
    val budget: TokenBudgetInfo
)

/**
 * Response from GET /chat/conversations/:id/messages.
 */
@Serializable
data class MessagesListResponse(
    val messages: List<Message>,
    val pagination: Pagination
)

/**
 * Generic status response for operations like archive.
 */
@Serializable
data class StatusResponse(
    val status: String,
    val message: String? = null
)

File: ios/Algo/Features/Chat/Models/ChatResponse.swift
// ChatResponse.swift
// Algo
//
// API response models for the chat endpoints, matching the backend API contract.
// Includes request models for sending messages and confirming actions.

import Foundation

// MARK: - ChatAPIResponse

/// Response from POST /chat/message — the full response envelope.
struct ChatAPIResponse: Codable, Sendable {

    let conversationId: String
    let messageId: String
    let response: ChatResponseBody
    let intent: IntentInfo?
    let reasoningSummary: String?
    let tokenBudget: TokenBudgetInfo?
    let timestamp: String

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case messageId = "message_id"
        case response
        case intent
        case reasoningSummary = "reasoning_summary"
        case tokenBudget = "token_budget"
        case timestamp
    }
}

// MARK: - ChatAPIResponse Convenience

extension ChatAPIResponse {

    /// Converts the API response into a `Message` model for display.
    func toMessage() -> Message {
        Message(
            id: messageId,
            role: .assistant,
            content: response.text,
            structuredData: response.structuredData,
            suggestedActions: response.suggestedActions,
            intent: intent?.classifiedAs,
            requiresConfirmation: response.requiresConfirmation,
            confirmationId: response.confirmationId,
            timestamp: parsedTimestamp
        )
    }

    /// Parses the ISO 8601 timestamp string into a `Date`.
    var parsedTimestamp: Date {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.date(from: timestamp) ?? Date()
    }
}

// MARK: - ChatResponseBody

/// The inner response body containing the assistant's reply content.
struct ChatResponseBody: Codable, Sendable {

    let text: String
    let structuredData: StructuredData?
    let suggestedActions: [SuggestedAction]?
    let requiresConfirmation: Bool
    let confirmationId: String?

    enum CodingKeys: String, CodingKey {
        case text
        case structuredData = "structured_data"
        case suggestedActions = "suggested_actions"
        case requiresConfirmation = "requires_confirmation"
        case confirmationId = "confirmation_id"
    }
}

// MARK: - IntentInfo

/// Classification result from the AI intent classifier.
struct IntentInfo: Codable, Sendable, Equatable {

    let classifiedAs: String
    let confidence: Double

    enum CodingKeys: String, CodingKey {
        case classifiedAs = "classified_as"
        case confidence
    }

    /// Whether the classification confidence meets the high-confidence threshold.
    var isHighConfidence: Bool {
        confidence >= 0.85
    }
}

// MARK: - TokenBudgetInfo

/// Token budget status included in chat responses for usage tracking.
struct TokenBudgetInfo: Codable, Sendable, Equatable {

    let remaining: Int
    let percentUsed: Int

    enum CodingKeys: String, CodingKey {
        case remaining
        case percentUsed = "percent_used"
    }

    /// Whether the budget is running low (>80% used).
    var isLow: Bool {
        percentUsed > 80
    }

    /// Whether the budget is critically low (>95% used).
    var isCritical: Bool {
        percentUsed > 95
    }
}

// MARK: - ChatRequest

/// Request body for POST /chat/message.
struct ChatRequest: Codable, Sendable {

    let conversationId: String?
    let message: String
    let inputType: String
    let audioUrl: String?
    let context: ChatContext?

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case message
        case inputType = "input_type"
        case audioUrl = "audio_url"
        case context
    }

    init(
        conversationId: String? = nil,
        message: String,
        inputType: String = "text",
        audioUrl: String? = nil,
        context: ChatContext? = nil
    ) {
        self.conversationId = conversationId
        self.message = message
        self.inputType = inputType
        self.audioUrl = audioUrl
        self.context = context
    }
}

// MARK: - ChatContext

/// Optional context about the user's current screen/selection sent with each message.
struct ChatContext: Codable, Sendable {

    let screen: String?
    let selectedCampaignId: String?

    enum CodingKeys: String, CodingKey {
        case screen
        case selectedCampaignId = "selected_campaign_id"
    }
}

// MARK: - ConfirmRequest

/// Request body for POST /chat/confirm.
struct ConfirmRequest: Codable, Sendable {

    let conversationId: String
    let confirmationId: String
    let confirmed: Bool

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case confirmationId = "confirmation_id"
        case confirmed
    }
}

// MARK: - MessageListResponse

/// Response from GET /chat/conversations/:id/messages.
struct MessageListResponse: Codable, Sendable {
    let messages: [Message]
    let pagination: Pagination
}

// MARK: - SuggestionsResponse

/// Response from GET /chat/suggestions — greeting suggestions for new conversations.
struct SuggestionsResponse: Codable, Sendable {

    let suggestions: [SuggestedAction]
    let greeting: String
    let timestamp: String
}

// MARK: - BudgetResponse

/// Response from GET /chat/budget — current token budget status and usage stats.
struct BudgetResponse: Codable, Sendable {

    let budget: BudgetDetails
    let usage: UsageStats?
    let timestamp: String
}

// MARK: - BudgetDetails

/// Detailed budget information from the token budget manager.
struct BudgetDetails: Codable, Sendable {

    let used: Int
    let limit: Int
    let remaining: Int
    let percentUsed: Int
    let isExhausted: Bool
    let resetInSeconds: Int

    enum CodingKeys: String, CodingKey {
        case used
        case limit
        case remaining
        case percentUsed = "percent_used"
        case isExhausted = "is_exhausted"
        case resetInSeconds = "reset_in_seconds"
    }
}

// MARK: - UsageStats

/// Accumulated usage statistics for the current tenant.
struct UsageStats: Codable, Sendable {

    let totalInputTokens: Int?
    let totalOutputTokens: Int?
    let requestCount: Int?

    enum CodingKeys: String, CodingKey {
        case totalInputTokens = "total_input_tokens"
        case totalOutputTokens = "total_output_tokens"
        case requestCount = "request_count"
    }
}
